docs/markdown/Using-wraptool.md

```diff
@@ -96,3 +96,16 @@ but available in WrapDB will automatically be downloaded.
 
 Automatic fetch of WrapDB subprojects can be disabled by removing the file
 `subprojects/wrapdb.json`, or by using `--wrap-mode=nodownload`.
+
+## Self-hosted Wrap database
+
+If for whatever reason you want to use self-hosted or proxied Wrap database *(since 1.X.X)* you may set mirror address to be used for the workspace:
+
+```console
+$ meson wrap set-db-mirror user:password@wrapdb.mydomain.invalid:8080
+$ meson wrap update-db
+$ meson wrap install zlib
+```
+
+You will be limited to the wraps available on the mirror, only one source is used at a time.
+The address is stored in `subprojects/wrapdb-mirrors.json`, remove the file to use upstream address.
```

Comment:
I am not sure I like the UI for this feature. I think that the use case you have in mind for it is very different from what I have in mind. Why isn't a command line argument to `meson wrap` enough? If this is a functionality intended mostly for corporate environment where access to the "upstream" wrap database is somehow forbidden, doesn't a network level redirect provide a transparent solution for the same problem?

Reply:
> May I propose an optional positional argument to the `update-db` command that will create the "persistent" file?

AFAIK running `meson wrap update-db` is not required to use other `meson wrap` commands. Therefore, making the configuration of the wrap database URL dependent on running it seems at least counterintuitive.

One thing I haven't understood yet is whether you want a setting that is per system, per user, per project, or per specific wrap. Depending on what is the use case the configuration would need to be stored in a different location.

I am also not convinced that there is the need of a meson command to set the WrapDB URL instead that a simple configuration file format.

Reply:
> You just need to change those measures to redirect access to that resource to https://wrapdb.localnet or whatever you like. Then you just serve your own wrap-db from that URL. Your wrap-db may contain your hand-picked selection of wraps. Why would this not work?

1. It will break TLS
2. Let's say that policy is only enforced on the CI machines
3. I have no control over this redirect
4. meson is overly attached to this specific address hardcoded in a few places

Reply:
> One thing I haven't understood yet is whether you want a setting that is per system, per user, per project, or per specific wrap. Depending on what is the use case the configuration would need to be stored in a different location.

Ideally setting wrapdb source per project with all of its subprojects. + A single instance of wrap server on the network.

> AFAIK running meson wrap update-db is not required to use other meson wrap commands. Therefore, making the configuration of the wrap database URL dependent on running it seems at least counterintuitive.

Certainly, it will not require `update-db` with the changes proposed in this MR. Only commands that query wrapdb need the update command whether with the source set or not. I don't think that one is supposed to have `wrapdb.json` file checked-in, and without it `status/info/search/list` won't work.

Reply:
> May I propose an optional positional argument to the update-db command that will create the "persistent" file?
I did it, but I don't like it. Will revert to a separate wrap subcommand.

Reply:
So far I find a file stored in subprojects to be the most convenient:

- does not depend on environment variables, thus cannot be easily hijacked by using environment variables
- has to be set once per project, as opposed to a `wrap` subcommands argument that will have to be set repeatedly when installing each wrap
- may be checked-in to the version control system, if project maintainer chooses to, eliminating the need for other developers to set it manually

Cons:
- it adds steps to the usual flow of `setup` + `compile`. Some build systems, that have wrappers around configuration scripts may need custom handling. Buildroot and Yocto come to mind, but these are bad examples, as both of them explicitly disable fallback and provide own mechanisms to deal with dependencies.